### ISSUE TITLE
consider adopting the kubectl "plugin" mechanism

### ISSUE THREAD

**Issue by starpit** [apache/incubator-openwhisk-client-go#110]
proposal: adopt kubectl's lightweight plugin mechanism for the openwhisk go cli. it's not even really a plugin mechanism, just a simple dispatcher. `kubectl foo` works if some executable `kubectl-foo` is on my PATH. kubectl is agnostic to the impl of kubectl-foo; i.e. it could be another go executable, a nodejs shebang, a bash script, etc.

getting this support into `wsk` should require only small changes, especially if we can leverage the platform's "which/type" support. i.e.: `wsk foo`, if not resolved by a built-in command or command subtree "foo" will delegate to the PATH resolver, which does a `which foo`. if this succeeds, it delegates the command line to an Exec of that resolved `/path/to/foo`.

**Comment by csantanapr** [comment]
I think this is a good thing to add
So then I can get “wsk deploy” if I rename wskdeploy to wsk-deploy !
cc @mrutkows @mdeuser

**Comment by pritidesai** [comment]
sounds like a great idea, @starpit what kind of changes are we anticipating? are you referring to something of this sort: https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/?

**Comment by mrutkows** [comment]
@starpit I am not opposed to this, but I still view this as a parallel path to what we want to do with making wskdeploy a plug-in of the CLI.  @pritidesai https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/

**Comment by starpit** [comment]
You are inventing a pkugin mechanism for wsk?

**Comment by mrutkows** [comment]
no, just advantaging existing Cobra framework

**Comment by starpit** [comment]
> sounds like a great idea, @starpit what kind of changes are we anticipating? are you referring to something of this sort: https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/?

@pritidesai yes, this is what i was referring to. it's an incredibly lightweight approach, and one that allows plugin providers to target multiple CLIs without having to invest anything into the possibly disparate mechanisms (ibmcloud, cobra, etc.) offered by each CLI provider.

**Comment by pritidesai** [comment]
thats great, have we experimented implementing this kind of setup outside of `kubectl`?

**Comment by starpit** [comment]
Not yet.
